### PR TITLE
brew.sh: fixes for UTF-8

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Fix Homebrew detection mechanism of usable locale.

Order of locales https://github.com/Homebrew/brew/pull/8047#issuecomment-662722092:

1. The user's current locale if it's UTF-8
2. C.UTF-8
3. en_US.UTF-8
4. First available UTF-8 locale
5. C

### CODE REVIEW

File: Library/Homebrew/brew.sh [hunk 1]
```diff
@@ -12,10 +12,28 @@ then
   then
     export LC_ALL="en_US.UTF-8"
   else
-    export LC_ALL="C.UTF-8"
+    utf8_locales=$(locale -a | grep -e \.utf8 -e \.UTF-8)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:15]:
```suggestion
    utf8_locales=$(locale -a | grep -iE '.utf-?8')`
```

File: Library/Homebrew/brew.sh [hunk 2]
```diff
@@ -12,10 +12,28 @@ then
...
+    if [[ -n $utf8_locales && 
+         ($utf8_locales =~ (^|[^a-zA-Z0-9_])(C|en_US)\.(utf8|UTF-8) ||
+          $utf8_locales =~ ^[a-zA-Z0-9_.]*) ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:18]: My understanding is the format is always `[a-z][a-z]_[A-Z][A-Z].(utf|UTF-)8` so this could be a bit stricter.

File: Library/Homebrew/brew.sh [hunk 3]
```diff
@@ -12,10 +12,28 @@ then
   then
     export LC_ALL="en_US.UTF-8"
   else
-    export LC_ALL="C.UTF-8"
+    utf8_locales=$(locale -a | grep -e \.utf8 -e \.UTF-8)
+    if [[ -n $utf8_locales && 
+         ($utf8_locales =~ (^|[^a-zA-Z0-9_])(C|en_US)\.(utf8|UTF-8) ||
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:17]: Could this be two separate `elif` statements instead of an `||` here?

File: Library/Homebrew/brew.sh [hunk 4]
```diff
@@ -12,10 +12,28 @@ then
...
+          $utf8_locales =~ ^[a-zA-Z0-9_.]*) ]]
+    then
+      export LC_ALL=$BASH_REMATCH
+    elif [[ -n $(locale -a | grep -w C) ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:21]:
```suggestion
    elif [[ -n $(locale -a | grep --word-regexp C) ]]
```

and as many other long flags as possible would make this a bit easier to follow.

File: Library/Homebrew/brew.sh [hunk 5]
```diff
@@ -12,10 +12,28 @@ then
...
+    then
+      export LC_ALL=C
+    else
+      echo "Error: Could not find usable locale" >&2
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:25]: It's not really an error if we continue as-is?

File: Library/Homebrew/brew.sh [hunk 6]
```diff
@@ -12,10 +12,28 @@ then
...
+    fi
   fi
 fi
 
+# Configure TTY to work with UTF8
+# https://unix.stackexchange.com/a/476671/222638
+if [[ $(stty -a | grep -o '.iutf8') == -iutf8 ]]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:32]: Why is this needed now and wasn't before?

**Comment by MikeMcQuaid** [comment]
`brew style` is failing due to `shellcheck`: https://github.com/Homebrew/brew/pull/8072/checks?check_run_id=904644217

File: Library/Homebrew/brew.sh (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:32]: Also `stty: 'standard input': Inappropriate ioctl for device` from the GitHub Actions output looks like a new error.

**Comment by maxim-belkin** [comment]
> `brew style` is failing due to `shellcheck`: https://github.com/Homebrew/brew/pull/8072/checks?check_run_id=904644217

Yeh, my bad -- I noticed that but had to leave and left the PR without a note that I'll continue working on it.
By the way, `brew style` doesn't fail for me in a Homebrew/brew Docker container, which is weird.

File: Library/Homebrew/brew.sh (hunk 1 above)
Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:15]:
```suggestion
    utf8_locales=$(locale -a | grep -E '\.(utf|UTF-)8')
```

File: Library/Homebrew/brew.sh (hunk 5 above)
Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:25]: Changing to "Warning"

File: Library/Homebrew/brew.sh (hunk 6 above)
Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:32]: Apparently, TTYs can support (or not) utf8 as well, so this step enables that support. I'm not sure this is necessary, so I'm going to remove it from this PR. We can look into this further if someone reports a problem.

**Comment by maxim-belkin** [comment]
Ready for a second review.

File: Library/Homebrew/brew.sh [hunk 7]
```diff
@@ -12,7 +12,19 @@ then
   then
     export LC_ALL="en_US.UTF-8"
   else
-    export LC_ALL="C.UTF-8"
+    utf8_locales=$(locale -a | grep --extended-regexp '\.(utf|UTF-)8')
+    re1='\bC\.(utf|UTF-)8'
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:16]: `c_regex`

File: Library/Homebrew/brew.sh [hunk 8]
```diff
@@ -12,7 +12,19 @@ then
...
+    re2='\ben_US\.(utf|UTF-)8'
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:17]: `en_us_regex`

File: Library/Homebrew/brew.sh [hunk 9]
```diff
@@ -12,7 +12,19 @@ then
...
+    re3='\b[a-z][a-z]_[A-Z][A-Z]\.(utf|UTF-)8'
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:18]: `utf_regex`

File: Library/Homebrew/brew.sh [hunk 10]
```diff
@@ -12,7 +12,19 @@ then
...
+    if [[ $utf8_locales =~ $re1 || $utf8_locales =~ $re2 || $utf8_locales =~ $re3 ]]
+    then
+      export LC_ALL=${BASH_REMATCH[0]}
+    elif locale -a | grep --quiet --word-regexp C
+    then
+      export LC_ALL=C
+    else
+      echo "Warning: Could not find usable locale." >&2
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:26]: What's the action item for the user here?

File: Library/Homebrew/brew.sh [hunk 11]
```diff
@@ -12,7 +12,19 @@ then
   then
     export LC_ALL="en_US.UTF-8"
   else
-    export LC_ALL="C.UTF-8"
+    utf8_locales=$(locale -a | grep --extended-regexp '\.(utf|UTF-)8')
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:15]: Would be good to store `locale -a` in a variable so it doesn't need re-called below. Also: could this use a Bash regex instead?

File: Library/Homebrew/brew.sh (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:16]: Why `\b`?

Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:16]: word boundary (on Linux platforms) to avoid matching things like `xx_XC.`. I don't know if there are any locales that match that but thought it wouldn't hurt to to be on a safe side. Should I remove it?

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:16]: Fine with me 👍🏻

File: Library/Homebrew/brew.sh (hunk 10 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:26]: (I'm wondering if this should be not output or be a hard-stop error).

Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:26]: install either a UTF-8 locale of choice or C locale.

Comment (MikeMcQuaid) [reply Library/Homebrew/brew.sh:26]: In that case, yeh, I'd make this an `odie` with instructions on how to resolve.

Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:24]: sounds good to me but I'd like to run this by @sjackman  👀

File: Library/Homebrew/brew.sh [hunk 12]
```diff
@@ -12,7 +12,17 @@ then
   then
     export LC_ALL="en_US.UTF-8"
   else
-    export LC_ALL="C.UTF-8"
+    locales=$(locale -a)
+    c_utf_regex='\bC\.(utf|UTF-)8'
+    en_us_regex='en_US\.(utf|UTF-)8'
+    utf_regex='[a-z][a-z]_[A-Z][A-Z]\.(utf|UTF-)8'
+    c_regex='\bC\b'
```
Comment (sjackman) [inline Library/Homebrew/brew.sh:19]:
```suggestion
    c_utf_regex='^C\.(utf8|UTF-8)$'
    en_us_regex='^en_US\.(utf8|UTF-8)$'
    utf_regex='^[a-z][a-z]_[A-Z][A-Z]\.(utf8|UTF-8)$'
    c_regex='^C$'
```
For style more than correctness. `^…$` is more common and easier to read than `\b`. Anchor all patterns for consistency. Super minor but I found `(utf8|UTF-8)` easier to read than `(utf|UTF-)8`.

File: Library/Homebrew/brew.sh (hunk 10 above)
Comment (sjackman) [reply Library/Homebrew/brew.sh:24]: The user is pretty much guaranteed to have a `C` locale. If they do not, the world is burning. It's more likely that `locale` failed to run (they don't have this executable in their `PATH`, which itself is pretty unlikely) than they don't have a `C` locale. Rather than emit a warning or error in this `else` case, I'd just set `LC_ALL=C` and move on.

**Comment by sjackman** [comment]
Thank you for this PR, Maxim!

File: Library/Homebrew/brew.sh (hunk 10 above)
Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:24]: > Rather than emit a warning or error in this `else` case, I'd just set `LC_ALL=C` and move on.

OK

File: Library/Homebrew/brew.sh (hunk 12 above)
Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:19]: `^...$` don't work as one may expect because `^` and `$` match the beginning and end of `locale -a` rather than the beginning and end of each word.

```sh
# In Homebrew/brew Docker image
$ locale -a
C
C.UTF-8
POSIX
en_US.utf8
$ locales=$(locale -a)
$ c_utf_regex='\bC\.(utf8|UTF-8)\b'
$ [[ $locales =~ $c_utf_regex ]] && echo ${BASH_REMATCH[0]}
C.UTF-8
$ c_utf_regex='^C\.(utf8|UTF-8)$'
$ [[ $locales =~ $c_utf_regex ]] && echo ${BASH_REMATCH[0]}
# nothing is printed; return status == 1
```

> Anchor all patterns for consistency.

OK to use `\b`?

>  Super minor but I found (utf8|UTF-8) easier to read than (utf|UTF-)8.

OK.

File: Library/Homebrew/brew.sh (hunk 10 above)
Comment (maxim-belkin) [reply Library/Homebrew/brew.sh:24]: Shall we remove `c_regex` then?

File: Library/Homebrew/brew.sh (hunk 12 above)
Comment (sjackman) [reply Library/Homebrew/brew.sh:19]: > OK to use \b?

Yep! Thanks for the explanation.

File: Library/Homebrew/brew.sh (hunk 10 above)
Comment (sjackman) [reply Library/Homebrew/brew.sh]: Makes sense.

**Comment by maxim-belkin** [comment]
Thanks for the reviews, @MikeMcQuaid and @sjackman!
